### DISCUSSION TITLE
add a new item to the menu bar that directly executes a command

### DISCUSSION THREAD

**Discussion by Adem-Kri** [eclipse-theia/theia#12402] [Q&A]
Hello,

I am building a Theia extension and have created a command that opens a new project. 

Currently, the command can only be accessed through the command palette. 

I would like to add a new item to the menu bar that directly executes the command when clicked. How can I achieve this in Theia?

P.S.: I am following the node-json-theia minimal example.

**Comment by vince-fugnitto** [comment]
@Adem-Kri thank you for the discussion, you can take a look at the documentation provided in the [website](https://theia-ide.org/docs/commands_keybindings/). We also rely heavily on the APIs in the framework if you need other [examples](https://github.com/eclipse-theia/theia/blob/master/examples/api-samples/src/browser/menu/sample-menu-contribution.ts).

**Reply by Adem-Kri** [reply]
Thank you for the reply @vince-fugnitto, 

I followed the documentation and I hadn't an output, can you please have a look  with me : 

So, I have this command code:
```ts
export const CREATE_PROJECT: Command = {
  id: "Cmd: roject-creator",
  label: "Cmd: New project creator ",
};
```
then, the class implementaion:
```ts
@injectable()
export class ProjectSetupContribution extends AbstractViewContribution<ProjectSetupWidget> {
  
constructor() {
    super({
      widgetId: ProjectSetupWidget.ID,
      widgetName: ProjectSetupWidget.LABEL,
      defaultWidgetOptions: { area: "main" },
    });
  }

  registerCommands(commands: CommandRegistry): void {
    commands.registerCommand(CREATE_PROJECT, {
      execute: () => {
        super
          .openView({ activate: false, reveal: true })
          .then(async (widget: ProjectSetupWidget) => {
            widget.open = true;
            widget.update();
          });
      },
    });
  }
}
```
this command works fine but when I implement the menu bar, the new item doesn't show up : 
```ts
@injectable()
export class MyProjectMenuContribution implements MenuContribution {
    registerMenus(menus: MenuModelRegistry): void {
        menus.registerMenuAction(MAIN_MENU_BAR, {
            commandId: CREATE_PROJECT.id,
            label: CREATE_PROJECT.label,
            order: "1",
        });
    }
}


export default new ContainerModule(bind => {
  bind(ProjectSetupContribution).toSelf();
  bind(MenuContribution).toDynamicValue(ctx => ctx.container.get(ProjectSetupContribution));
});
```

I'll be grateful for any suggestions, thank you.

**Reply by vince-fugnitto** [reply]
@Adem-Kri are you able to share your extension in a public github repository to get the full picture?

**Reply by Adem-Kri** [reply]
@vince-fugnitto, sorry but i am not allowed to do that.

**Reply by vince-fugnitto** [reply]
I think your binding might be incorrect? It seems as though you are using the wrong symbol `ProjectSetupContribution` > `MyProjectMenuContribution`?

```ts
export default new ContainerModule(bind => {
  bind(MyProjectMenuContribution).toSelf();
  bind(MenuContribution).toService(MyProjectMenuContribution);
});

**Reply by Adem-Kri** [reply]
@vince-fugnitto also didn't work, I just changed the naming because I am not to share functions and methods names, can you tell depending on my code what I am supposed to find as result, I mean the item name or the position maybe it is up there and I didn't notice it

**Reply by vince-fugnitto** [reply]
The following code:

```ts
menus.registerMenuAction(MAIN_MENU_BAR, {
    commandId: CREATE_PROJECT.id,
    label: CREATE_PROJECT.label,
    order: "1",
});
```

Should add an entry `Cmd: New project creator` at the topmost level of the main-menu. If you are still experiencing issues please take a look at the dev console to see if there are any errors.
